### ISSUE TITLE
Help in understanding TVMWrap_GetInputPtr and TVMWrap_GetOutputPtr

### ISSUE THREAD

**Issue by idealbuq** [tum-ei-eda/utvm_staticrt_codegen#14]
Hello, 
First of all I'd like to thank you, it is indeed a great work what you've done.
I have a question not really an issue, but I am having trouble understanding the call to _TVMWrap_GetInputPtr_ and _TVMWrap_GetOutputPtr_ functions in _main.c_. 
For example, for the given model example _cifar10.tflite,_ which has an input of size (32x32x3) how come we only pass a float (3.14/2f) to TVMWrap_GetInputPtr(0) ? And same for TVMWrap_GetOutputPtr, I'm not sure about what "result" actually represents.
Thanks in advance ! 

```
int main()
{
    TVMWrap_Init();

    *(float*)TVMWrap_GetInputPtr(0) = 3.14/2f;

    TVMWrap_Run();

    float result;
    *(volatile float*)&result = *(float *)TVMWrap_GetOutputPtr(0);

    DBGPRINTF("out: %f\n", result);
}
```

**Comment by rafzi** [comment]
Hello, thanks for you interest!

This is actually just a lazy test that does not do anything useful for cifar10. The test code was primarility intended for the sine_model.tflite, where we pass a value x and expect ~sin(x) in the output. It is also "valid" for cifar10, but the computation is not meaningful.

For cifar10, the input is supposed to be an image from the "cifar10" dataset (3 channels for 3 colors) and the output is the softmax of the 10 classes that you can classify with this dataset. The GetInput/OutputPtr functions just return a base pointer where you can copy the flattened tensor data. For writing a bit more generic code, you have `TVMWrap_GetInputSize`, `TVMWrap_GetOutputSize`, `TVMWrap_GetNumInputs` and `TVMWrap_GetNumOutputs`, but usually you'd know your model inputs and outputs anyway.

**Comment by idealbuq** [comment]
Thank you for your reply!
Ok, that's what I thought. 
I tried passing a flattened tensor but I don't think I am doing this correctly as it doesn't work, what am I missing?
Should I pass it in the runtime_wrapper.c instead?
That's what I tried: 

```
int main()
{
    TVMWrap_Init();

    float input_data[32 * 32 * 3];
    FILE* fp = fopen("cat.bin", "rb");
    (void)fread(input_data, 3072, 4, fp);
    fclose(fp);

    *(float*)TVMWrap_GetInputPtr(0) = *(float*)&input_data;

    TVMWrap_Run();

    float result[10];
    *(volatile float*)&result = *(float *)TVMWrap_GetOutputPtr(0);
    
    for(int i = 0; i < 10; i++) { 
        DBGPRINTF("out: %f\n", result[i]);
    }
}
```
Thanks in advance :)

**Comment by rafzi** [comment]
Alright, I tried out the model and it seems to work fine for me.

Based on this example, the input preparation is a bit more complex than just flattening: https://github.com/deep-diver/CIFAR10-img-classification-tensorflow/blob/master/CIFAR10_image_classification.py

So I used the following code for preparing some test data. The input file is the "python version" of the dataset.

```
import struct
import pickle


with open("data_batch_1", "rb") as f:
    batch = pickle.load(f, encoding="latin1")

features = batch['data'].reshape((len(batch['data']), 3, 32, 32)).transpose(0, 2, 3, 1)
labels = batch['labels']

for n in range(0, 10):
    data = b""
    for i in range(0, 32*32*3):
        v = features[n].flatten()[i]
        v = float(v) / 255
        data += struct.pack("f", v)

    with open("input/" + str(n) + ".bin", "wb") as f:
        f.write(data)

    print("Label", labels[n])
```

If I use the resulting binary files as input, I get the following output for the first three images, which matches the labels from the dataset:

    -5.70180511,-7.33687449,1.12904537,1.30579066,0.636840641,1.99929345,3.69460416,3.66799283,-5.66724348,-7.26373482
    -> highest: 6
    0.55146718,-2.15028167,-0.291549176,-2.43002295,-3.20001435,-2.59800267,-7.58031082,3.3427918,0.470968604,8.39957428
    -> highest 9
    2.75353885,1.3455565,-1.51044416,-1.64057207,-3.36897111,-2.75683594,-4.08160257,3.24108791,-2.57346535,5.26013899
    -> highest 9

Small correction: It is not using Softmax, but just outputting ten categories where the highest value represents the prediction.

**Comment by idealbuq** [comment]
Ok, thank you for the explanation of the CIFAR dataset preprocessing.
But is what I showed you previously the correct way to pass the input data to TVMWrap_GetInputPtr(0)? Or maybe I am not knowing how to read the output from TVMWrap_GetOutputPtr(0).
If you could help me with the calling of those functions in main.c I'd be glad. Thank you very much for your attention !
Again I tried: 

```
int main()
{
    TVMWrap_Init();

    float input_data[32 * 32 * 3];
    FILE* fp = fopen("input/0.bin", "r");
    (void)fread(input_data, 3072, 4, fp);
    fclose(fp);

    *(float*)TVMWrap_GetInputPtr(0) = *( float*)&input_data;

    TVMWrap_Run();

    float result[10];
    *( float*)&result = *(float *)TVMWrap_GetOutputPtr(0);
    
    for(int i = 0; i < 10; i++) { 
        DBGPRINTF("out: %f\n", result[i]);
    }
}
```

**Comment by rafzi** [comment]
No, your code is not correct.

TVMWrap_GetInput/OutputPtr returns a pointer to the start of the buffer.

Your line

    *(float*)TVMWrap_GetInputPtr(0) = *( float*)&input_data;

copies the address of input_data (casted as float) to the first float of the buffer. What you probably want is:

    memcpy(TVMWrap_GetInputPtr(0), input_data, sizeof(input_data));

For the output, this might be most practical:

    float result[10];
    for(int i = 0; i < 10; i++) { 
        result[i] = ((float*)TVMWrap_GetOutputPtr(0))[i];
        DBGPRINTF("out: %f\n", result[i]);
    }

**Comment by idealbuq** [comment]
Yes, it works like a charm! Thank you very much for helping me understand this.
